docs/src/main/asciidoc/security-openid-connect-client.adoc

```diff
@@ -0,0 +1,218 @@
+
+== Use OidcClient directly
+
+One can use `OidcClient` directly as follows:
+
+```
+import javax.inject.PostConstruct;
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+
+import io.quarkus.oidc.client.OidcClient;
+import io.quarkus.oidc.client.GrantTokens;
+
+@Path("/service")
+public class OidcClientResource {
+
+    @Inject
+    OidcClient client;
+
+    volatile GrantTokens currentTokens;
+
+    @PostConstruct
+    public init() {
+        currentTokens = client.getTokens().await().indefinitely();
+    }
+
+    @GET
+    public String getResponse() {
+        
+        GrantTokens tokens = currentTokens;
+        if (tokens.isAccessTokenExpired) {
+            tokens = client.refreshTokens(tokens.getRefreshToken().await.indefinitely();
+            currentTokens = tokens;
+        } 
+        // use tokens.getAccessToken() to configure MP RestClient Authorization header/etc
+    }
+}
+```
+
+== Use OidcClient as Microprofile RestClient client filter
+
+`quarkus-oidc-rest-client` extension provides `io.quarkus.oidc.rest.client.OidcClientRequestFilter` JAX-RS ClientRequestFilter which uses `OidcClient` to acquire the access token, refresh it if needed, and set it as an HTTP `Authorization` `Bearer` scheme value.
+
+By default this filter will get `OidcClient` to acquire the first pair of access and refresh tokens at its initialization time. If the access tokens are short lived and refresh tokens are not available then the token acquisition should be delayed with `quarkus.oidc-client.early-tokens-acquisition=false`.
+
+You can selectively register `OidcClientRequestFilter` by using either `io.quarkus.oidc.rest.client.OidcClientFilter` or `org.eclipse.microprofile.rest.client.annotation.RegisterProvider` annotations:
+
+```
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+import io.quarkus.oidc.rest.client.OidcClientFilter;
+
+@RegisterRestClient
+@OidcClientFilter
+@Path("/")
+public interface ProtectedResourceService {
+
+    @GET
+    String getUserName();
+}
+```
+
+or
+
+```
+import org.eclipse.microprofile.rest.client.annotation.RegisterProvider;
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+import io.quarkus.oidc.rest.client.OidcClientRequestFilter;
+
+@RegisterRestClient
+@RegisterProvider(OidcClientRequestFilter.class)
+@Path("/")
+public interface ProtectedResourceService {
+
+    @GET
+    String getUserName();
+}
+```
+
+Alternatively, `OidcClientRequestFilter` can be registered automatically with all MP Rest or JAX-RS clients if `quarkus.oidc.rest.client.register-filter=true` property is set. 
+
+
+== Use injected GrantTokens
+
+If you prefer you can use your own custom filter and inject `GrantTokens`:
+
+```
+@Provider
+@Priority(Priorities.AUTHENTICATION)
+@RequestScoped
+public class OidcClientRequestCustomFilter implements ClientRequestFilter {
+
+    @Inject
+    GrantTokens grantTokens;
```

Comment:
I don't think you should allow direct injection of the GrantTokens, as there is no CDI scope that this maps to. It would be very easy to accidentally inject them into a long lived object, which will appear to work for a while until the tokens expire (e.g. likely work in tests and staging, then fail once the production env has been up for a while). This example only works because you have made the filter @RequestScoped, which is not something we really encourage.

I think you should instead inject a GrantTokensProducer, which as a get() and getAsync() method. This object can be application scoped and actually manage the tokens in a thread safe manner so that new tokens don't need to be acquired per request.

Reply:
@stuartwdouglas As I mentioned above I have a working test - but I'll review that test, may be it is broken :-)

Reply:
@stuartwdouglas Hi Stuart, I've resolved 2 related comments, thanks for highlighting the problems with this specific producer; so, first of all `GrantTokens` and `GrantTokensProducer` got renamed based on @pedroigor 's suggestion to simplify so the `Grant` part has gone from their names.
`TokensProducer` is now `@Singleton` but it `@RequestScoped Produces` the tokens. Because it is a singleton it gets the initial set of tokens on its initialization time only once and then it will either return them or refresh them. This should be correct now.
The other thing about `TokensProducer` and the `injected Tokens` is that it is not a main thing this PR is after - `OidcClient` already provides the direct async access to `Tokens`, and `OidcClientRequestFilter` manages them too which is what I'd expect most users would use, `injected Tokens` is a little utility for those users who just don't want to use the provided filter or `OidcClient` request directly and it now works correctly I believe.

By the way, I've added a check log code (where the log is checked that it contains a refresh token confirmation) but what I've found is that apparently, when multiple tests run, as in `oidc-client/deployment`, the `target` dir seems to be locked somehow, and this check log only works if I run the individual test,
i.e, the following works:
```
mvn install -Pdocker-keycloak -Ptest-keycloak -Dtest=OidcClientUserPasswordCustomFilterTestCase
```
This test would successfully detect `target/quarkus.log` and check the records there.
The same test fails with:
```
mvn install -Pdocker-keycloak -Ptest-keycloak
```
It runs more than one test, not only `OidcClientUserPasswordCustomFilterTestCase`, and `OidcClientUserPasswordCustomFilterTestCase` fails with the message that `target/quarkus.log` is not available.

Do you know what may be going on ? I've removed the checkLog code from it for now, as `integration-tests/oidc-client` also checks the log and it works fine

Reply:
@stuartwdouglas I've also updated `integration-tests/oidc-client` to check that log only contains a single acquisition and refresh entry, there are a few tests there which depend on a singleton `OidcClientRequestFilter`, before I added `@Singleton` to it it re-acquiring for every test so yeah now it is all good I think